### PR TITLE
Add experimental HTTP/3 support to client and the server

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

**Comment by wendigo** [comment]
@dain I'd like to experiment with HTTP/3 in Trino. Since this is an opt-in feature (for both client/server) we could potentially get this in.

File: http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java [hunk 1]
```diff
@@ -141,6 +146,34 @@ public HttpClientConfig setHttp2Enabled(boolean http2Enabled)
...
+    @Config("http-client.http3.pem-path")
+    @ConfigDescription("Enable the experimental HTTP/3 transport")
+    public HttpClientConfig setHttp3PemPath(String http3PemPath)
+    {
+        this.http3PemPath = Optional.ofNullable(http3PemPath)
+                .map(Path::of);
+        return this;
+    }
+
+    public boolean isHttp3Enabled()
```
Comment (electrum) [inline http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java:172]: Wrong order, this should be above. The order should be getter+setter

File: http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java [hunk 2]
```diff
@@ -141,6 +146,34 @@ public HttpClientConfig setHttp2Enabled(boolean http2Enabled)
...
 
+    @Config("http-client.http3.enabled")
+    @ConfigDescription("Enable the experimental HTTP/3 transport")
+    public HttpClientConfig setHttp3Enabled(boolean http3Enabled)
+    {
+        this.http3Enabled = http3Enabled;
+        return this;
+    }
+
+    @ConfigDescription("Path that will be used to export PEM certificates to use for HTTP/3")
```
Comment (electrum) [inline http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java:157]: This should be moved to the setter for `setHttp3PemPath`

File: http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java [hunk 3]
```diff
@@ -437,9 +440,18 @@ protected void configure(SelectableChannel selectable)
...
-    private HttpClientTransport getClientTransport(ClientConnector connector, HttpClientConfig config)
+    private HttpClientTransport getClientTransport(ClientConnector connector, SslContextFactory.Client sslContextFactory, HttpClientConfig config)
     {
         ImmutableList.Builder<ClientConnectionFactory.Info> protocols = ImmutableList.builder();
+
+        if (config.isHttp3Enabled()) {
+            ClientQuicConfiguration clientQuicConfig = new ClientQuicConfiguration(sslContextFactory, config.getHttp3PemPath()
+                    .orElseThrow(() -> new RuntimeException("http-client.http3.pem-path needs to be set when http-client.http3.enabled")));
+            HTTP3Client http3Client = new HTTP3Client(clientQuicConfig);
+            http3Client.getQuicConfiguration().setSessionRecvWindow(64 * 1024 * 1024);
```
Comment (electrum) [inline http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java:451]: Is this how much memory the client will use? If so, might want to make this configurable

File: http-client/src/test/java/io/airlift/http/client/TestingHttpServer.java [hunk 4]
```diff
@@ -74,16 +88,38 @@ public TestingHttpServer(Optional<String> keystore, Servlet servlet, Consumer<Ht
...
+                    httpConfiguration.addCustomizer(new SvcResponseCustomizer(port));
+                    // Configure the HTTP3 connector port to be the same as HTTPS/HTTP2.
+                    quicServerConnector.setPort(port);
+                }
+            });
         }
         else {
             connector = new ServerConnector(server, insecureFactories(httpConfiguration));
+            configureConnector(scheme, connector);
+            connector.setReusePort(true);
```
Comment (electrum) [inline http-client/src/test/java/io/airlift/http/client/TestingHttpServer.java:119]: Is this needed for non-QUIC?

File: http-client/src/test/java/io/airlift/http/client/jetty/TestJettyHttpClientHttp3.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+public class TestJettyHttpClientHttp3
+        extends AbstractHttpClientTest
+{
+    private Path tempPemDir;
+
+    public TestJettyHttpClientHttp3()
+            throws IOException
+    {
+        super(getResource("localhost.keystore").toString());
+        tempPemDir = Files.createTempDirectory("pem");
```
Comment (electrum) [inline http-client/src/test/java/io/airlift/http/client/jetty/TestJettyHttpClientHttp3.java:25]: Clean this up at the end of the test?

**Review by electrum — APPROVED** [review]

**Pull request closed without merging**